File: hardware/scala/nafarr/system/timer/TimerCtrl.scala
// SPDX-FileCopyrightText: 2026 aesc silicon
//
// SPDX-License-Identifier: CERN-OHL-W-2.0

package nafarr.system.timer

import spinal.core._
import spinal.lib._
import spinal.lib.bus.misc.BusSlaveFactory
import spinal.lib.misc.InterruptCtrl

import nafarr.IpIdentification

object TimerCtrl {
  def apply(p: Parameter) = new TimerCtrl(p)

  case class Parameter(
      count: Int = 1,
      channelCount: Int = 1,
      width: Int = 32,
      prescalerWidth: Int = 16
  ) {
    require(count >= 1 && count <= 16, "count must be 1..16")
    require(channelCount >= 1 && channelCount <= 8, "channelCount must be 1..8")
    require(width >= 1 && width <= 32, "width must be 1..32")
    require(prescalerWidth >= 0 && prescalerWidth <= 32, "prescalerWidth must be 0..32")
    val timerStride: Int = 0x10 + channelCount * 0x04
    val irqSources: Int = count * (1 + channelCount)
  }
  object Parameter {
    def default() = Parameter()
    def small() = Parameter(count = 1, channelCount = 1, width = 16, prescalerWidth = 8)
    def medium() = Parameter(count = 1, channelCount = 2, width = 32, prescalerWidth = 16)
    def large() = Parameter(count = 4, channelCount = 2, width = 32, prescalerWidth = 16)
  }

  object Regs {
    def apply(base: BigInt, p: Parameter) = new Regs(base, p)
  }
  class Regs(base: BigInt, p: Parameter) {
    val info = base + 0x00
    val irqPending = base + 0x04
    val irqMask = base + 0x08
    private def timerBase(t: Int): BigInt = base + 0x0c + t * p.timerStride
    def control(t: Int) = timerBase(t) + 0x00
    def prescaler(t: Int) = timerBase(t) + 0x04
    def counter(t: Int) = timerBase(t) + 0x08
    def reload(t: Int) = timerBase(t) + 0x0c
    def compare(t: Int, ch: Int) = timerBase(t) + 0x10 + ch * 0x04
  }

  /** Per-timer configuration driven from the Mapper into the controller. */
  case class TimerInstanceConfig(p: Parameter) extends Bundle {
    val enable = Bool()
    val mode = Bits(2 bits) // 00=free-run 01=periodic 10=one-shot
    val reloadVal = UInt(p.width bits)
    val compareVals = Vec(UInt(p.width bits), p.channelCount)
  }

  /** Per-timer status driven from the controller back to the Mapper. */
  case class TimerInstanceStatus(p: Parameter) extends Bundle {
    val counter = UInt(p.width bits)
    val overflow = Bool()
    val compareMatch = Vec(Bool(), p.channelCount)
    val enableClear = Bool() // one-shot done: Mapper must clear the enable bit
  }

  case class Io(p: Parameter) extends Bundle {
    val config = in Vec (TimerInstanceConfig(p), p.count)
    val counterLoad = in Vec (Bool(), p.count)
    val counterIn = in Vec (UInt(p.width bits), p.count)
    val status = out Vec (TimerInstanceStatus(p), p.count)
    val prescaler = (p.prescalerWidth > 0) generate in(Vec(UInt(p.prescalerWidth bits), p.count))
  }

  /** Timer controller.
    *
    * Contains all counter and prescaler hardware. No bus logic.
    *
    * config       : per-timer enable, mode, reload, compare values (from Mapper registers).
    * counterLoad  : preload strobe - loads counterIn into the counter on the next edge.
    *               Hardware update has priority; preload only takes effect when stopped.
    * counterIn    : preload value written by software.
    * prescaler    : prescaler reload value (only present when prescalerWidth > 0).
    *
    * status.counter      : current counter value (readable by Mapper via bus).
    * status.overflow     : pulses one cycle on counter bound (all modes).
    * status.compareMatch : pulses one cycle when counter equals compareVals[ch].
    * status.enableClear  : pulses one cycle on completion of a one-shot run.
    *
    * Modes (mode[1:0]):
    *   00 free-run  - counts 0 to maxVal, wraps to 0.
    *   01 periodic  - counts 0 to reloadVal, resets to 0; repeats.
    *   10 one-shot  - counts 0 to reloadVal, resets to 0, asserts enableClear.
    */
  case class TimerCtrl(p: Parameter) extends Component {
    val io = Io(p)

    for (t <- 0 until p.count) {
      new Area {
        val cfg = io.config(t)
        val sts = io.status(t)

        // Prescaler: counts down from prescalerReg; generates tick on zero.
        val tick = Bool()
        if (p.prescalerWidth > 0) {
          val prescalerCounter = Reg(UInt(p.prescalerWidth bits)) init (0)
          tick := False
          when(prescalerCounter === 0) {
            tick := True
            prescalerCounter := io.prescaler(t)
          } otherwise {
            prescalerCounter := prescalerCounter - 1
          }
        } else {
          tick := True
        }

        // Counter register.
        val counterReg = Reg(UInt(p.width bits)) init (0)
        sts.counter := counterReg

        // Overflow: at maxVal for free-run, at reloadVal for periodic/one-shot.
        val maxVal = U((BigInt(1) << p.width) - 1, p.width bits)
        val atBound = Bool()
        when(cfg.mode === B"00") {
          atBound := counterReg === maxVal
        } otherwise {
          atBound := counterReg === cfg.reloadVal
        }

        sts.overflow := cfg.enable && tick && atBound
        sts.enableClear := cfg.enable && tick && atBound && (cfg.mode === B"10")

        // Compare matches (checked against current counter value each tick).
        for (ch <- 0 until p.channelCount) {
          sts.compareMatch(ch) := cfg.enable && tick && (counterReg === cfg.compareVals(ch))
        }

        // Counter update. Hardware has priority; preload only takes effect when stopped.
        when(cfg.enable && tick) {
          when(atBound) {
            counterReg := 0
          } otherwise {
            counterReg := counterReg + 1
          }
        } elsewhen (io.counterLoad(t)) {
          counterReg := io.counterIn(t)
        }
      }.setName(s"timer_$t")
    }
  }

  case class Mapper(
      busCtrl: BusSlaveFactory,
      ctrl: TimerCtrl,
      p: Parameter
  ) extends Area {
    val idCtrl = IpIdentification(IpIdentification.Ids.Timer, 1, 0, 0)
    idCtrl.driveFrom(busCtrl)
    val regs = Regs(idCtrl.length, p)

    // info: [7:0]=count [15:8]=channelCount [23:16]=width [31:24]=prescalerWidth
    busCtrl.read(
      B(p.prescalerWidth, 8 bits) ## B(p.width, 8 bits) ##
        B(p.channelCount, 8 bits) ## B(p.count, 8 bits),
      regs.info
    )

    // InterruptCtrl accumulates events; io.masks hardwired to all-ones so io.pendings
    // returns raw (unmasked) pending on bus reads. A separate maskReg drives the
    // interrupt output, keeping standard W1C-at-irqPending / mask-at-irqMask layout.
    val irqCtrl = new InterruptCtrl(p.irqSources)
    irqCtrl.io.masks := B((BigInt(1) << p.irqSources) - 1, p.irqSources bits)
    val clearFlow = busCtrl.createAndDriveFlow(Bits(p.irqSources bits), regs.irqPending.toInt)
    irqCtrl.io.clears := 0
    when(clearFlow.valid) { irqCtrl.io.clears := clearFlow.payload }
    busCtrl.read(irqCtrl.io.pendings.resized, regs.irqPending.toInt)
    val maskReg = Reg(Bits(p.irqSources bits)) init (0)
    busCtrl.readAndWrite(maskReg, regs.irqMask.toInt)
    val irqEvents = Vec(Bool(), p.irqSources)

    // Prescaler registers (only when prescalerWidth > 0).
    if (p.prescalerWidth > 0) {
      for (t <- 0 until p.count) {
        val prescalerReg = Reg(UInt(p.prescalerWidth bits)) init (0)
        busCtrl.readAndWrite(prescalerReg, regs.prescaler(t))
        ctrl.io.prescaler(t) := prescalerReg
      }
    } else {
      for (t <- 0 until p.count) {
        busCtrl.read(B(0, 32 bits), regs.prescaler(t))
      }
    }

    for (t <- 0 until p.count) {
      new Area {
        // Control register: hardware clears enable on one-shot completion.
        val controlFlow = busCtrl.createAndDriveFlow(Bits(3 bits), regs.control(t))
        val controlReg = Reg(Bits(3 bits)) init (0)
        when(ctrl.io.status(t).enableClear) {
          controlReg := B"000"
        } elsewhen (controlFlow.valid) {
          controlReg := controlFlow.payload
        }
        busCtrl.read(controlReg, regs.control(t))
        ctrl.io.config(t).enable := controlReg(0)
        ctrl.io.config(t).mode := controlReg(2 downto 1)

        // Reload register.
        val reloadReg = Reg(UInt(p.width bits)) init (0)
        busCtrl.readAndWrite(reloadReg, regs.reload(t))
        ctrl.io.config(t).reloadVal := reloadReg

        // Compare registers.
        for (ch <- 0 until p.channelCount) {
          val compareReg =
            Reg(UInt(p.width bits)) init (U((BigInt(1) << p.width) - 1, p.width bits))
          busCtrl.readAndWrite(compareReg, regs.compare(t, ch))
          ctrl.io.config(t).compareVals(ch) := compareReg
        }

        // Counter: bus reads live value; bus write preloads (when stopped).
        val counterFlow = busCtrl.createAndDriveFlow(UInt(p.width bits), regs.counter(t))
        ctrl.io.counterLoad(t) := counterFlow.valid
        ctrl.io.counterIn(t) := counterFlow.payload
        busCtrl.read(ctrl.io.status(t).counter, regs.counter(t))

        // Interrupt sources: overflow at irqBase, compare[ch] at irqBase+1+ch.
        val irqBase = t * (1 + p.channelCount)
        irqEvents(irqBase) := ctrl.io.status(t).overflow
        for (ch <- 0 until p.channelCount) {
          irqEvents(irqBase + 1 + ch) := ctrl.io.status(t).compareMatch(ch)
        }
      }.setName(s"timer_$t")
    }

    irqCtrl.io.inputs := irqEvents.asBits
    val interrupt = (irqCtrl.io.pendings & maskReg).orR
  }
}

File: hardware/scala/nafarr/system/timer/TimerCtrl.cs
// SPDX-FileCopyrightText: 2026 aesc silicon
//
// SPDX-License-Identifier: CERN-OHL-W-2.0
//
// Renode functional model of the Nafarr general-purpose timer (nafarr.system.timer.TimerCtrl),
// for ElemRV-N's single-timer / single-channel / 16-bit / 8-bit-prescaler configuration
// (TimerCtrl.Parameter.small()). Register-accurate; the counter advances in virtual time and
// raises IRQ on compare/overflow so timer-driven firmware runs. It is NOT a co-simulated RTL model
// (the kernel tick uses the machine timer); it exists so the general-purpose timer can stay
// functional and out of the co-simulation for speed. Wire IRQ to its PLIC source, e.g. `-> plic@2`.
//
// Register map (offsets from base; regOffset = IpIdentification.length = 8):
//   0x00 header    (RO)  IpIdentification (api<<24 | length<<16 | id=23) = 0x00080017
//   0x04 version   (RO)  1.0.0
//   0x08 info      (RO)  prescalerWidth<<24 | width<<16 | channelCount<<8 | count = 0x08100101
//   0x0C irqPending(W1C)  [0]=overflow [1]=compare-match; write-1-to-clear
//   0x10 irqMask   (RW)  output-IRQ enable per source
//   0x14 control   (RW)  [0]=enable [2:1]=mode (00=free-run 01=periodic 10=one-shot)
//   0x18 prescaler (RW)  clock divider (tick every prescaler+1)
//   0x1C counter   (RW)  read live value; write preloads while stopped
//   0x20 reload    (RW)  reload value (periodic)
//   0x24 compare   (RW)  channel-0 compare value (period in periodic/one-shot)
//
using Antmicro.Renode.Core;
using Antmicro.Renode.Logging;
using Antmicro.Renode.Peripherals.Bus;
using Antmicro.Renode.Peripherals.Timers;
using Antmicro.Renode.Time; // Direction, WorkMode

namespace Antmicro.Renode.Peripherals.Timers
{
    public class TimerCtrl : IDoubleWordPeripheral, IKnownSize
    {
        public TimerCtrl(IMachine machine, long frequency = 30000000)
        {
            IRQ = new GPIO();
            timer = new LimitTimer(machine.ClockSource, (ulong)frequency, this, "timer",
                                   limit: CounterMax, direction: Direction.Ascending,
                                   enabled: false, workMode: WorkMode.Periodic,
                                   eventEnabled: true, autoUpdate: true);
            timer.LimitReached += OnLimitReached;
            Reset();
        }

        public void Reset()
        {
            control = 0;
            prescaler = 0;
            reload = 0;
            compare = 0;
            pending = 0;
            mask = 0;
            timer.Enabled = false;
            timer.Value = 0;
            UpdateTimer();
            IRQ.Set(false);
        }

        public uint ReadDoubleWord(long offset)
        {
            switch(offset)
            {
                case HeaderOffset:     return Header;
                case VersionOffset:    return 0x01000000; // 1.0.0
                case InfoOffset:       return Info;
                case IrqPendingOffset: return pending;
                case IrqMaskOffset:    return mask;
                case ControlOffset:    return control;
                case PrescalerOffset:  return prescaler;
                case CounterOffset:    return (uint)(timer.Value & CounterMask);
                case ReloadOffset:     return reload;
                case CompareOffset:    return compare;
                default:
                    this.Log(LogLevel.Warning, "Unhandled read at offset 0x{0:X}", offset);
                    return 0;
            }
        }

        public void WriteDoubleWord(long offset, uint value)
        {
            switch(offset)
            {
                case IrqPendingOffset:  // write-1-to-clear
                    pending &= ~(value & 0x3u);
                    UpdateIrq();
                    return;
                case IrqMaskOffset:
                    mask = value & 0x3u;
                    UpdateIrq();
                    return;
                case ControlOffset:
                    control = value & 0x7u;
                    UpdateTimer();
                    return;
                case PrescalerOffset:
                    prescaler = value & 0xFFu;
                    UpdateTimer();
                    return;
                case CounterOffset:     // preload only takes effect while stopped (matches RTL)
                    if(!Enabled)
                    {
                        timer.Value = value & CounterMask;
                    }
                    return;
                case ReloadOffset:
                    reload = value & (uint)CounterMask;
                    return;
                case CompareOffset:
                    compare = value & (uint)CounterMask;
                    UpdateTimer();
                    return;
                case HeaderOffset:
                case VersionOffset:
                case InfoOffset:
                    this.Log(LogLevel.Warning, "Write to read-only register at offset 0x{0:X}", offset);
                    return;
                default:
                    this.Log(LogLevel.Warning, "Unhandled write at offset 0x{0:X}", offset);
                    return;
            }
        }

        public long Size => 0x1000;
        public GPIO IRQ { get; }

        private bool Enabled => (control & 0x1u) != 0;
        private int Mode => (int)((control >> 1) & 0x3u); // 0=free-run 1=periodic 2=one-shot

        private void UpdateTimer()
        {
            // Period: compare (periodic/one-shot); full range for free-run or an unset compare.
            timer.Limit = (Mode == 0 || compare == 0) ? CounterMax : compare;
            timer.Divider = (ulong)(prescaler + 1);
            timer.Mode = (Mode == 2) ? WorkMode.OneShot : WorkMode.Periodic;
            timer.Enabled = Enabled;
        }

        private void OnLimitReached()
        {
            // Free-run -> overflow (bit 0); periodic/one-shot -> compare match (bit 1).
            pending |= (Mode == 0) ? 0x1u : 0x2u;
            if(Mode == 2) // one-shot: hardware clears the enable bit on completion
            {
                control &= ~0x1u;
                timer.Enabled = false;
            }
            UpdateIrq();
        }

        private void UpdateIrq()
        {
            IRQ.Set((pending & mask) != 0);
        }

        private uint control;
        private uint prescaler;
        private uint reload;
        private uint compare;
        private uint pending;
        private uint mask;
        private readonly LimitTimer timer;

        private const ulong CounterMask = 0xFFFF; // width = 16
        private const ulong CounterMax = 0xFFFF;
        // IpIdentification: id = Timer = 23, length = 8, api = 0.
        private const uint Header = (0u << 24) | (8u << 16) | 23u;
        // info: prescalerWidth=8, width=16, channelCount=1, count=1.
        private const uint Info = (8u << 24) | (16u << 16) | (1u << 8) | 1u;

        private const long HeaderOffset = 0x00;
        private const long VersionOffset = 0x04;
        private const long InfoOffset = 0x08;
        private const long IrqPendingOffset = 0x0C;
        private const long IrqMaskOffset = 0x10;
        private const long ControlOffset = 0x14;
        private const long PrescalerOffset = 0x18;
        private const long CounterOffset = 0x1C;
        private const long ReloadOffset = 0x20;
        private const long CompareOffset = 0x24;
    }
}
